WIP: Support for readonly on jakarta persistence

OmarHawk opened the pull request:
Contains changes to more easily extend jakarta persistence configuration in `@Column` definition

#27048

<!--
PR description.
-->

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

<!--
Please also reference the issue number in a commit message to [automatically close the related GitHub issue](https://help.github.com/articles/closing-issues-via-commit-messages/)

Note: It is also possible to add `[skip ci]` or `[ci skip]` to your commit message to skip continuous integration tests
-->

OmarHawk committed:
Refactor to avoid duplicate code

and to more easily extend jakarta persistence configuration in Column

OmarHawk committed:
Support for readonly in backend side

Fix #27048

OmarHawk commented:
By the way, for some reasons, it complains about some Sonar findings - but the only thing I changed there was to extend one test entity. I do wonder, if I need to/should do something about it in this Pull request

mshima commented:
New entities added new code and increased complexity, the sonar error is not blocking.

OmarHawk commented:
Ok, and any hints regarding testability of these added attributes, if anything that happens in our own generated code won't notice any change due to Spring Data JPA's behaviour?

mshima commented:
@OmarHawk readonly is quite uncommon use case.
This looks related to https://github.com/jhipster/generator-jhipster/issues/25414
Use saveAndFlush for jpa and entity with fields that requires flush like `@AutoGenerated`?

Not sure it’s possible to test a pure `@Readonly` field.
Maybe another approach is required for created timestamp.
Add a `@AutoGenerated` which implies `@Readonly` annotation and test if the returned is not null?

mshima commented:
Add readonly to typescript type?

OmarHawk commented:
I guess, I may finish this in around 4 weeks - heading off to some vacation. :-) 

I think, the very first commit here would already be "worth" integrating, as it does only refactor the template logic for the `@Column` annotation parameters, but doesn't change anything to the generated applications.
The readonly support may be "extra" then.

mraible commented:
@OmarHawk Any chance you want to look at this before tomorrow's release?

OmarHawk commented:
No, won't manage that. Actually I even forgot about this here, but I'd take it up in the next few days then.

OmarHawk committed:
Load persisted entity again if there are readonly fields

to return the actual state of the entity and not something set from user side. May break tests now.

OmarHawk commented:
> @OmarHawk readonly is quite uncommon use case. This looks related to #25414 Use saveAndFlush for jpa and entity with fields that requires flush like `@AutoGenerated`?
> 
> Not sure it’s possible to test a pure `@Readonly` field. Maybe another approach is required for created timestamp. Add a `@AutoGenerated` which implies `@Readonly` annotation and test if the returned is not null?

I added now something that reloads the entity state from database if there are readonly fields in the entity  - I restricted it to these entities to only do that when it is really necessary:
Only this way we can be sure to return the actual value, as the `insertable = false` and `updateable = false` are only a hint to not include these columns in a INSERT / UPDATE command. So the return value of the EntityRepository methods would always return the user set values. The only way to get the actual values are to load these again from database.

Some tests are supposed to fail now, as I haven't updated the tests now to reflect this change.

DanielFran commented:
@OmarHawk Can you please update your branch, this can go with v9 ;)

OmarHawk commented:
Sure, ignoring is correct here, as this is nothing I can fix

OmarHawk commented:
Btw. sorry, for taking so long, but currently super busy with work related stuff, so tbh, this has not the highest prio. :-(

Currently still open is especially the reactive stack + test adjustments.

OmarHawk commented:
yes, I will have to rebase or start over again and remove some of the commits as these changes have already been integrated in main (especially the small refactoring part within this current branch)
